AzureMonitor: add NewDimension component using experimental UI

asimpson opened the pull request:
**What this PR does / why we need it**:
This wraps up the Azure Monitor metrics UI migration to use the experimental UI (ref #44432) by creating a new Dimensions component. 

A key part of this is using [`EditorList`](https://github.com/grafana/grafana-experimental/blob/main/src/query-builder/EditorList.tsx) for the Dimension filters. `EditorList` handles manipulation of the data structure passed via the `items` prop so we can remove a few helper functions such as `addFilter`, `removeFilter`, and helper functions for `onChange`.

This new component is exercised by the same unit test file as the current Dimension component.

**Which issue(s) this PR fixes**:

Closes #48709

asimpson committed:
AzureMonitor: add NewDimension component using experimental UI

This new component is exercised by the same unit test file as the current Dimension component.

Also cleans up a few unneeded `await` keywords in the Dimensions test file.

kevinwcyu commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/NewDimensionFields.tsx:
```diff
@@ -0,0 +1,207 @@
+  };
+
+  return (
+    <Field label="Dimensions">
```
Should this be imported from `public/app/plugins/datasource/grafana-azure-monitor-datasource/components/Field.tsx` instead of `@grafana/ui`.

asimpson replied:
Good catch!

sarahzinger commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/NewDimensionFields.tsx:
```diff
@@ -0,0 +1,207 @@
+        .filter((item) => item!);
+      for (const label of labels) {
+        // Labels only exist for series that have a dimension selected
+        if (label) {
```
it's kind of weird that you have to check that the label exists right because didn't we already filter out the false-y ones on line 31?

sarahzinger commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/NewDimensionFields.tsx:
```diff
@@ -0,0 +1,207 @@
+        return prevLabels;
+      }
+      for (const label of currentLabels) {
+        if (prevLabels[label] && labelsObj[label].size < prevLabels[label].size) {
```
hey sorry do you mind explaining what is going on here? why are we only setting the bigger sized label

aangelisc replied:
I can explain this - we use the bigger sized label to ensure that we don't remove any of the potential dimension options. When the user chooses a dimension option the series will change and filter out the ones that don't match the condition. To ensure users can easily switch between dimensions we choose the larger labels object which will be unfiltered😊

sarahzinger commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/NewDimensionFields.tsx:
```diff
@@ -0,0 +1,207 @@
+    value: AzureMetricDimension[Key],
+    onChange: (item: Partial<AzureMetricDimension>) => void
+  ) => {
+    item[fieldName] = value;
```
does this mutate the original item? Is that a bit weird?

sarahzinger reviewed (approved):
seems ok to me! 🎉 I had some questions/thoughts but no real issues

asimpson committed:
AzureMonitor: make tweaks based on PR comments.

- I was importing the wrong Field component
- We can use a typeguard to avoid the strange `if`.

asimpson commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.test.tsx:
```diff
@@ -18,342 +19,365 @@ const variableOptionGroup = {
 
-describe('Azure Monitor QueryEditor', () => {
-  const mockDatasource = createMockDatasource();
+const tests = [
```
Totally agree that the array approach doesn't feel as comfortable here as it does in Go. However, I think it's worth the tradeoff here because it lets us verify _every_ test with the feature flag on and the feature flag off without changing the existing tests _too_ much. Also, if we eventually don't need the feature flag I'd be all for removing the array pattern here as part of the feature flag removal. TLDR;  the array pattern is very much tied to the feature flag.
